### PR TITLE
First stab at relative delta functionality.

### PR DESCRIPTION
**Pull request by vext01**
This is my first attempt at relative deltas in warmup stats.

We add the ability to pass `--delta X%`, where the delta is X% of the *fastest* segment.

This is not a codebase I am familiar with, so please check carefully.

I found two places where we have to compute the absolute delta.

For now this is a draft PR.

### CODE REVIEW

**Comment by vext01** [comment]
Ah travis. I didn't realise we had tests. Maybe it will find some failures for us.

**Comment by vext01** [comment]
I don't understand:
```
$ ./test/test.sh

Marking outliers with sliding window size: 200

Loading: test/example1.json.bz2

Writing out: test/example1_outliers_w200.json.bz2

Marking outliers with sliding window size: 200

Loading: test/example2.json.bz2

Writing out: test/example2_outliers_w200.json.bz2

Traceback (most recent call last):

  File "./bin/mark_changepoints_in_json", line 46, in <module>

    from warmup.statistics import get_absolute_delta_using_fastest_seg

ImportError: No module named warmup.statistics
```

This works locally when I run `test.sh`.

Here's the module it claims doesn't exist:
https://github.com/vext01/warmup_stats/blob/percent-delta/warmup/statistics.py

File: warmup/statistics.py [hunk 1]
```diff
@@ -70,3 +70,20 @@ def bootstrap_runner(marshalled_data, quality='HIGH'):
...
+
+
+def get_absolute_delta_using_fastest_seg(delta, seg_means):
+    return get_absolute_delta(delta, min(seg_means))
+
+
+def get_absolute_delta(delta, seg_time):
+    assert type(delta) in (str, unicode)
+
+    try:
```
Comment (ltratt) [inline warmup/statistics.py]: Might be simpler as:

```python
if  delta.endswith("%"):
   return seg_time * (float(delta[:-1]) / 100)
else:
    return float(delta)
```

**Comment by ltratt** [comment]
OK, so this looks good, and I don't think needs much further work.

One thing that I'd forgotten is that the numbers we're getting are not "what happens if the tolerance is X%" but "what happens if the tolerance is X% or the variance of the last segment, whichever is greater"?

**Comment by vext01** [comment]
> OK, so this looks good, and I don't think needs much further work.

Cool. I think we need to update the help text though.

> One thing that I'd forgotten is that the numbers we're getting are not "what happens if the tolerance is X%" but "what happens if the tolerance is X% or the variance of the last segment, whichever is greater"?

What I thought I had implemented here was: "what happens if the tolerance is X% of the fastest segment in the process execution". I'll have a look tomorrow to see if the variance (you mean CIs, right?) form some part of the decision.

**Comment by ltratt** [comment]
You can see it here on line 148(ish) e.g.:

```python
        lower_bound = min(last_segment.mean - last_segment.variance,
                          last_segment.mean - self.delta)
```

**Comment by vext01** [comment]
Hrm. I'm not 100% sure what we want actually. Any thoughts.

I want to know if the variance really is the variance or if it is actually a CI. I've recently learned that the variance is problematic.

**Comment by ltratt** [comment]
We talk about why we use the variance (and the trade-offs of doing so) in the paper (p11). My guess is that it makes very little difference except at the very LHS of the graph (which is why the % plot starts with a little flat section before becoming logarithmic). We can disable it if we want, but we can equally well explain it.

**Comment by vext01** [comment]
The problem is, the variance isn't measured in (in this case) seconds. But it sounds like we knew that:

> variance... (we simply interpreted it as having units in seconds rather than seconds squared).

I don't think that's strictly correct, but at least it answers my question.

**Comment by ltratt** [comment]
That's why we said that we used it is a heuristic for what we wanted, so it's not exactly an issue of correctness.

**Comment by vext01** [comment]
OK, but does it undermine what we are doing here with %age bounds?

**Comment by ltratt** [comment]
I think that for the graphs we want to draw in the paper we might want to avoid the variance entirely. Not because it's the variance (it's a perfectly fine, if unusual, heuristic), but because it slightly interferes with what we're trying to show. Whatever heuristic we use for "allow a little tolerance" there will be times when it overtakes the delta, and the latter is what we concentrate on. So we might want some way of switching the variance heuristic off (either as a temporary hack or as an explicit flag).

**Comment by vext01** [comment]
I agree. I think the heuristic isn't suitable for what we are trying to show.

Will fix this tomorrow. Thanks

**Comment by vext01** [comment]
The last commit implements a flag `--raw-deltas` which ignores the variance in the computation previously discussed.

Shall we re-run "quick plots" to see what difference it makes before we merge this?

File: bin/mark_changepoints_in_json [hunk 2]
```diff
@@ -149,10 +151,16 @@ class Segments(object):
         delta = get_absolute_delta_using_fastest_seg(
             self.delta, [s.mean for s in self.segments])
 
-        lower_bound = min(last_segment.mean - last_segment.variance,
-                          last_segment.mean - delta)
-        upper_bound = max(last_segment.mean + last_segment.variance,
-                          last_segment.mean + delta)
+        if self.raw_deltas:
+            import pdb; pdb.set_trace()
```
Comment (ltratt) [inline bin/mark_changepoints_in_json]: We need to remove this line.

**Comment by ltratt** [comment]
> Shall we re-run "quick plots" to see what difference it makes before we merge this?

Yes, probably. The commit looks decent (minus the one comment), but checking is definitely a good idea!

File: bin/mark_changepoints_in_json (hunk 2 above)
Comment (vext01) [reply bin/mark_changepoints_in_json]: haha, oops!

**Comment by ltratt** [comment]
I think this is ready for squashing.

**Comment by vext01** [comment]
Pushed a change to a argument doc. Looks OK?

**Comment by ltratt** [comment]
Works for me. Please squash.

**Comment by vext01** [comment]
splat

**Pull request merged**